File: src/main/java/destiny/astrology/classical/rules/debilities/Detriment.java
/**
 * @author smallufo
 * Created on 2007/12/30 at 上午 4:46:40
 */
package destiny.astrology.classical.rules.debilities;

import destiny.astrology.Horoscope;
import destiny.astrology.Planet;
import destiny.astrology.classical.Dignity;
import org.jetbrains.annotations.NotNull;
import org.jooq.lambda.tuple.Tuple;
import org.jooq.lambda.tuple.Tuple2;

import java.util.Optional;

/** In Detriment. */
public final class Detriment extends EssentialRule {

  public Detriment() {
  }

  @Override
  protected Optional<Tuple2<String, Object[]>> getResult(@NotNull Planet planet, @NotNull Horoscope h) {
    return h.getZodiacSign(planet)
      .filter(sign -> planet == essentialImpl.getPointOptional(sign , Dignity.DETRIMENT).orElse(null))
      .map(sign -> Tuple.tuple("comment", new Object[]{planet, sign}));
  }

}

File: src/main/java/destiny/astrology/classical/rules/debilities/Detriment.kt
/**
 * @author smallufo
 * Created on 2007/12/30 at 上午 4:46:40
 */
package destiny.astrology.classical.rules.debilities

import destiny.astrology.IHoroscopeModel
import destiny.astrology.Planet

/** In Detriment.  */
class Detriment : EssentialRule() {

  override fun getResult(planet: Planet, h: IHoroscopeModel): Pair<String, Array<Any>>? {

    return h.getZodiacSign(planet)
      ?.takeIf { sign -> planet === with(detrimentImpl) { sign.getDetrimentPoint() } }
      ?.let { "comment" to arrayOf(planet , it) }
  }
}
